File: gatling-http-client/src/main/java/io/gatling/http/client/realm/DigestAuth.java
/*
 * Copyright 2011-2021 GatlingCorp (https://gatling.io)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package io.gatling.http.client.realm;

import io.gatling.netty.util.StringBuilderPool;
import io.netty.handler.codec.http.HttpMethod;

import java.security.MessageDigest;
import java.util.concurrent.ThreadLocalRandom;

import static io.gatling.http.client.util.MessageDigestUtils.pooledMd5MessageDigest;
import static io.gatling.http.client.util.StringUtils.*;
import static java.nio.charset.StandardCharsets.ISO_8859_1;
import static io.gatling.http.client.util.MiscUtils.*;

class DigestAuth {

  // MD5("")
  private static final String EMPTY_ENTITY_MD5 = "d41d8cd98f00b204e9800998ecf8427e";

  private final String username;
  private final String password;
  private final String realm;
  private final String nonce;
  private final String opaque;
  private final String algorithm;
  private final String qop;
  private final String nc;
  private final HttpMethod requestMethod;
  private final String realmUri;


  DigestAuth(String username,
             String password,
             String realm,
             String nonce,
             String opaque,
             String algorithm,
             String qop,
             String nc,
             HttpMethod requestMethod,
             String realmUri) {
    this.username = username;
    this.password = password;
    this.realm = realm;
    this.nonce = nonce;
    this.opaque = opaque;
    this.algorithm = algorithm;
    this.qop = qop;
    this.nc = nc;
    this.requestMethod = requestMethod;
    this.realmUri = realmUri;
  }

  public String computeAuthorization() {
    MessageDigest md = pooledMd5MessageDigest();
    String cnonce = newCnonce(md);
    String response = newResponse(md, cnonce);

    StringBuilder builder = new StringBuilder().append("Digest ");
    append(builder, "username", username, true);
    append(builder, "realm", realm, true);
    append(builder, "nonce", nonce, true);
    append(builder, "uri", realmUri, true);
    if (isNonEmpty(algorithm)) {
      append(builder, "algorithm", algorithm, false);
    }
    append(builder, "response", response, true);

    if (opaque != null) {
      append(builder, "opaque", opaque, true);
    }

    if (qop != null) {
      append(builder, "qop", qop, false);
      // nc and cnonce only sent if server sent qop
      append(builder, "nc", nc, false);
      append(builder, "cnonce", cnonce, true);
    }
    builder.setLength(builder.length() - 2); // remove tailing ", "

    // FIXME isn't there a more efficient way?
    return new String(charSequence2Bytes(builder, ISO_8859_1));
  }

  private static String newCnonce(MessageDigest md) {
    byte[] b = new byte[8];
    ThreadLocalRandom.current().nextBytes(b);
    b = md.digest(b);
    return toHexString(b);
  }

  private String newResponse(MessageDigest md, String cnonce) {

    StringBuilder sb = StringBuilderPool.DEFAULT.get();

    // WARNING: DON'T MOVE, BUFFER IS RECYCLED!!!!
    byte[] ha1 = ha1(sb, md, cnonce);
    byte[] ha2 = ha2(sb, realmUri, md);

    appendBase16(sb, ha1);
    appendMiddlePart(sb, cnonce);
    appendBase16(sb, ha2);

    byte[] responseDigest = md5FromRecycledStringBuilder(sb, md);
    return toHexString(responseDigest);
  }

  private byte[] md5FromRecycledStringBuilder(StringBuilder sb, MessageDigest md) {
    md.update(charSequence2ByteBuffer(sb, ISO_8859_1));
    sb.setLength(0);
    return md.digest();
  }

  private byte[] ha1(StringBuilder sb, MessageDigest md, String cnonce) {
    // if algorithm is "MD5" or is unspecified => A1 = username ":" realm-value ":" passwd
    // if algorithm is "MD5-sess" => A1 = MD5( username-value ":" realm-value ":" passwd ) ":" nonce-value ":" cnonce-value

    sb.append(username).append(':').append(realm).append(':').append(password);
    byte[] core = md5FromRecycledStringBuilder(sb, md);

    if (algorithm == null || algorithm.equals("MD5")) {
      // A1 = username ":" realm-value ":" passwd
      return core;
    } else if ("MD5-sess".equals(algorithm)) {
      // A1 = MD5(username ":" realm-value ":" passwd ) ":" nonce ":" cnonce
      appendBase16(sb, core);
      sb.append(':').append(nonce).append(':').append(cnonce);
      return md5FromRecycledStringBuilder(sb, md);
    }

    throw new UnsupportedOperationException("Digest algorithm not supported: " + algorithm);
  }

  private byte[] ha2(StringBuilder sb, String digestUri, MessageDigest md) {

    // if qop is "auth" or is unspecified => A2 = Method ":" digest-uri-value
    // if qop is "auth-int" => A2 = Method ":" digest-uri-value ":" H(entity-body)
    sb.append(requestMethod.name()).append(':').append(digestUri);
    if ("auth-int".equals(qop)) {
      // when qop == "auth-int", A2 = Method ":" digest-uri-value ":" H(entity-body)
      // but we don't have the request body here
      // we would need a new API
      sb.append(':').append(EMPTY_ENTITY_MD5);

    } else if (qop != null && !qop.equals("auth")) {
      throw new UnsupportedOperationException("Digest qop not supported: " + qop);
    }

    return md5FromRecycledStringBuilder(sb, md);
  }

  private void appendMiddlePart(StringBuilder sb, String cnonce) {
    // request-digest = MD5(H(A1) ":" nonce ":" nc ":" cnonce ":" qop ":" H(A2))
    sb.append(':').append(nonce).append(':');
    if ("auth".equals(qop) || "auth-int".equals(qop)) {
      sb.append(nc).append(':').append(cnonce).append(':').append(qop).append(':');
    }
  }

  private static void append(StringBuilder builder, String name, String value, boolean quoted) {
    builder.append(name).append('=');
    if (quoted)
      builder.append('"').append(value).append('"');
    else
      builder.append(value);

    builder.append(", ");
  }
}

File: gatling-http/src/main/scala/io/gatling/http/auth/DigestAuth.scala
/*
 * Copyright 2011-2022 GatlingCorp (https://gatling.io)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package io.gatling.http.auth

import java.nio.charset.StandardCharsets.UTF_8
import java.security.MessageDigest
import java.util.Base64
import java.util.concurrent.ThreadLocalRandom

import io.gatling.commons.util.StringHelper.RichString
import io.gatling.commons.validation.Validation
import io.gatling.http.client.uri.Uri
import io.gatling.http.client.util.StringUtils.toHexString
import io.gatling.netty.util.StringBuilderPool

import io.netty.handler.codec.http.HttpMethod

object DigestAuth {

  private val ParserPool = new ThreadLocal[DigestWwwAuthenticateHeaderParser] {
    override def initialValue: DigestWwwAuthenticateHeaderParser = new DigestWwwAuthenticateHeaderParser
  }

  private val Md5DigestPool = new ThreadLocal[MessageDigest] {
    override def initialValue: MessageDigest = MessageDigest.getInstance("MD5")
  }

  private val Sha256DigestPool = new ThreadLocal[MessageDigest] {
    override def initialValue: MessageDigest = MessageDigest.getInstance("SHA-256")
  }

  private val Sha512256DigestPool = new ThreadLocal[MessageDigest] {
    override def initialValue: MessageDigest = MessageDigest.getInstance("SHA-512/256")
  }

  private val SbPool = new StringBuilderPool

  def parseWwwAuthenticateHeader(header: String, origin: Uri): Validation[Challenge] =
    ParserPool.get().parse(header, origin)

  private def newCnonce(): String = {
    val b = new Array[Byte](8)
    ThreadLocalRandom.current.nextBytes(b)
    Base64.getEncoder.encodeToString(b)
  }

  def hash(s: String, md: MessageDigest): String = toHexString(md.digest(s.getBytes(UTF_8)))

  def generateAuthorization(challenge: Challenge, username: String, password: String, requestMethod: HttpMethod, requestUri: Uri, nc: Int): String = {
    generateAuthorization0(challenge, username, password, requestMethod, requestUri, nc, newCnonce())
  }

  // for test only
  private[auth] def generateAuthorization0(
      challenge: Challenge,
      username: String,
      password: String,
      requestMethod: HttpMethod,
      requestUri: Uri,
      nc: Int,
      cnonce: String
  ): String = {
    val algorithm = challenge.algorithm
    val md = algorithm match {
      case Algorithm.Md5 | Algorithm.Md5Sess             => Md5DigestPool.get()
      case Algorithm.Sha256 | Algorithm.Sha256Sess       => Sha256DigestPool.get()
      case Algorithm.Sha512256 | Algorithm.Sha512256Sess => Sha512256DigestPool.get()
    }
    val qop = challenge.qop

    val hashedUsername = if (challenge.userhash) hash(s"$username:${challenge.realm}", md) else username
    val paddedNc = nc.toString.leftPad(8, "0")

    val a1 =
      if (challenge.algorithm.session) {
        // if -sess algo:     A1 = H( unq(username) ":" unq(realm) ":" passwd ) ":" unq(nonce-prime) ":" unq(cnonce-prime)
        s"${hash(s"$username:${challenge.realm}:$password", md)}:${challenge.nonce}:$cnonce"
      } else {
        // if standard algo:  A1 = unq(username) ":" unq(realm) ":" passwd
        s"$username:${challenge.realm}:$password"
      }

    // we don't support "auth-int" for now
    // if qop=auth        A2       = Method ":" request-uri
    // if qop=auth-int    A2       = Method ":" request-uri ":" H(entity-body)
    val a2 = s"${requestMethod.name}:${requestUri.toRelativeUrl}"

    // H(concat(H(A1), ":", unq(nonce), ":", nc, ":", unq(cnonce), ":", unq(qop), ":", H(A2)))
    val response = hash(s"${hash(a1, md)}:${challenge.nonce}:$paddedNc:$cnonce:${qop.value}:${hash(a2, md)}", md)

    val sb = SbPool
      .get()
      .append("Digest ")
      .append("username=\"")
      .append(hashedUsername)
      .append("\", realm=\"")
      .append(challenge.realm)
      .append("\", uri=\"")
      .append(requestUri.toRelativeUrl)
      .append("\", algorithm=")
      .append(algorithm.value)
      .append(", nonce=\"")
      .append(challenge.nonce)
      .append("\", nc=")
      .append(paddedNc)
      .append(", cnonce=\"")
      .append(cnonce)
      .append("\", qop=")
      .append(qop.value)
      .append(", response=\"")
      .append(response)
      .append("\"")

    challenge.opaque.foreach(opaque => sb.append(", opaque=\"").append(opaque).append("\""))

    if (challenge.userhash) {
      sb.append(", userhash=true")
    }

    sb.toString
  }

  sealed abstract class Qop(val value: String)

  object Qop {
    case object Auth extends Qop("auth")
    case object AuthInt extends Qop("auth-int")
  }

  sealed abstract class Algorithm(val value: String, val session: Boolean, val securityLevel: Int) extends Product with Serializable

  object Algorithm {
    case object Md5 extends Algorithm("MD5", session = false, 0)
    case object Md5Sess extends Algorithm("MD5-sess", session = true, 0)
    case object Sha256 extends Algorithm("SHA-256", session = false, 1)
    case object Sha256Sess extends Algorithm("SHA-256-sess", session = true, 1)
    case object Sha512256 extends Algorithm("SHA-512-256", session = false, 2)
    case object Sha512256Sess extends Algorithm("SHA-512-256-sess", session = true, 2)
  }

  final case class ProtectionSpace(domain: String, path: String) {
    def matches(uri: Uri): Boolean = uri.getHost == domain && uri.getNonEmptyPath.startsWith(path)
  }

  final case class Challenge(
      realm: String,
      domain: Set[ProtectionSpace],
      nonce: String,
      opaque: Option[String],
      stale: Boolean,
      algorithm: Algorithm,
      qop: Qop,
      userhash: Boolean
  )
}

final case class DigestAuth() {}
